src/imp/linux_raw/arch/inline/aarch64.rs

```diff
@@ -21,259 +21,265 @@ pub(in crate::imp) unsafe fn syscall0_readonly(nr: SyscallNumber<'_>) -> RetReg<
 ) -> RetReg<R0> {
     let r0;
     asm!(
         "svc 0",
         in("x8") nr.to_asm(),
-        inlateout("x0") a0.to_asm() => r0,
+        inlateout("x0") a0.into().to_asm() => r0,
         options(nostack, preserves_flags, readonly)
     );
     FromAsm::from_asm(r0)
 }
 
 #[inline]
 #[must_use]
-pub(in crate::imp) unsafe fn syscall1_noreturn(nr: SyscallNumber<'_>, a0: ArgReg<'_, A0>) -> ! {
+pub(in crate::imp) unsafe fn syscall1_noreturn<'a>(
+    nr: SyscallNumber<'_>,
+    a0: impl Into<ArgReg<'a, A0>>,
+) -> ! {
     asm!(
         "svc 0",
         in("x8") nr.to_asm(),
-        in("x0") a0.to_asm(),
+        in("x0") a0.into().to_asm(),
         options(noreturn)
     )
 }
 
 #[inline]
 #[must_use]
-pub(in crate::imp) unsafe fn syscall2(
+pub(in crate::imp) unsafe fn syscall2<'a>(
     nr: SyscallNumber<'_>,
-    a0: ArgReg<'_, A0>,
-    a1: ArgReg<'_, A1>,
+    a0: impl Into<ArgReg<'a, A0>>,
+    a1: impl Into<ArgReg<'a, A1>>,
 ) -> RetReg<R0> {
     let r0;
     asm!(
         "svc 0",
         in("x8") nr.to_asm(),
-        inlateout("x0") a0.to_asm() => r0,
-        in("x1") a1.to_asm(),
+        inlateout("x0") a0.into().to_asm() => r0,
+        in("x1") a1.into().to_asm(),
         options(nostack, preserves_flags)
     );
     FromAsm::from_asm(r0)
 }
 
 #[inline]
 #[must_use]
-pub(in crate::imp) unsafe fn syscall2_readonly(
+pub(in crate::imp) unsafe fn syscall2_readonly<'a>(
     nr: SyscallNumber<'_>,
-    a0: ArgReg<'_, A0>,
-    a1: ArgReg<'_, A1>,
+    a0: impl Into<ArgReg<'a, A0>>,
+    a1: impl Into<ArgReg<'a, A1>>,
 ) -> RetReg<R0> {
     let r0;
     asm!(
         "svc 0",
         in("x8") nr.to_asm(),
-        inlateout("x0") a0.to_asm() => r0,
-        in("x1") a1.to_asm(),
+        inlateout("x0") a0.into().to_asm() => r0,
+        in("x1") a1.into().to_asm(),
         options(nostack, preserves_flags, readonly)
     );
     FromAsm::from_asm(r0)
 }
 
 #[inline]
 #[must_use]
-pub(in crate::imp) unsafe fn syscall3(
+pub(in crate::imp) unsafe fn syscall3<'a>(
     nr: SyscallNumber<'_>,
-    a0: ArgReg<'_, A0>,
-    a1: ArgReg<'_, A1>,
-    a2: ArgReg<'_, A2>,
+    a0: impl Into<ArgReg<'a, A0>>,
+    a1: impl Into<ArgReg<'a, A1>>,
+    a2: impl Into<ArgReg<'a, A2>>,
 ) -> RetReg<R0> {
     let r0;
     asm!(
         "svc 0",
         in("x8") nr.to_asm(),
-        inlateout("x0") a0.to_asm() => r0,
-        in("x1") a1.to_asm(),
-        in("x2") a2.to_asm(),
+        inlateout("x0") a0.into().to_asm() => r0,
+        in("x1") a1.into().to_asm(),
+        in("x2") a2.into().to_asm(),
         options(nostack, preserves_flags)
     );
     FromAsm::from_asm(r0)
 }
 
 #[inline]
 #[must_use]
-pub(in crate::imp) unsafe fn syscall3_readonly(
+pub(in crate::imp) unsafe fn syscall3_readonly<'a>(
     nr: SyscallNumber<'_>,
-    a0: ArgReg<'_, A0>,
-    a1: ArgReg<'_, A1>,
-    a2: ArgReg<'_, A2>,
+    a0: impl Into<ArgReg<'a, A0>>,
+    a1: impl Into<ArgReg<'a, A1>>,
+    a2: impl Into<ArgReg<'a, A2>>,
```

Comment:
Any concerns about the implicit generics here potentially introducing multiple copies of this and other functions?  Might we need to do the "manual forward to non-generic impl" trick?  (Seems like that'd be painful here though...)

Or maybe the inline asm is small enough this doesn't matter.

Reply:
Ah, that's a good point. In release builds these I expect these to be inlined. But I should look more closely at what happens in practice, and also what happens in debug builds.

Reply:
Probably simplest to `#[inline(always)]`?

Reply:
It's a good idea, but thinking about this more, I think this is a good use case for a few macros. They can help factor out even more of the redundancy, including a bunch of `#cfg`'s in the syscalls.rs files, which is quite nice and appears enough for me to set aside my general dislike of macros in this case :-).